pkg/apiserver/registry/core/namespacedcloudprofile/strategy.go

```diff
@@ -0,0 +1,118 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package namespacedcloudprofile
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/storage/names"
+
+	"github.com/gardener/gardener/pkg/api"
+	"github.com/gardener/gardener/pkg/apis/core"
+	"github.com/gardener/gardener/pkg/apis/core/validation"
+)
+
+type namespacedCloudProfileStrategy struct {
+	runtime.ObjectTyper
+	names.NameGenerator
+}
+
+// Strategy defines the storage strategy for NamespacedCloudProfiles.
+var Strategy = namespacedCloudProfileStrategy{api.Scheme, names.SimpleNameGenerator}
+
+func (namespacedCloudProfileStrategy) NamespaceScoped() bool {
+	return true
+}
+
+func (namespacedCloudProfileStrategy) PrepareForCreate(_ context.Context, obj runtime.Object) {
+	namespacedCloudProfile := obj.(*core.NamespacedCloudProfile)
+
+	dropExpiredVersions(namespacedCloudProfile)
+	namespacedCloudProfile.Status = core.NamespacedCloudProfileStatus{}
+}
+
+func (namespacedCloudProfileStrategy) Validate(_ context.Context, obj runtime.Object) field.ErrorList {
+	namespacedCloudProfile := obj.(*core.NamespacedCloudProfile)
+	return validation.ValidateNamespacedCloudProfile(namespacedCloudProfile)
+}
+
+func (namespacedCloudProfileStrategy) Canonicalize(_ runtime.Object) {
+}
+
+func (namespacedCloudProfileStrategy) AllowCreateOnUpdate() bool {
+	return false
+}
+
+func (namespacedCloudProfileStrategy) PrepareForUpdate(_ context.Context, newObj, oldObj runtime.Object) {
+	_ = oldObj.(*core.NamespacedCloudProfile)
+	_ = newObj.(*core.NamespacedCloudProfile)
+}
+
+func (namespacedCloudProfileStrategy) AllowUnconditionalUpdate() bool {
+	return true
+}
+
+func (namespacedCloudProfileStrategy) ValidateUpdate(_ context.Context, newObj, oldObj runtime.Object) field.ErrorList {
+	oldProfile, newProfile := oldObj.(*core.NamespacedCloudProfile), newObj.(*core.NamespacedCloudProfile)
+	return validation.ValidateNamespacedCloudProfileUpdate(newProfile, oldProfile)
+}
+
+// WarningsOnCreate returns warnings to the client performing the create.
+func (namespacedCloudProfileStrategy) WarningsOnCreate(_ context.Context, _ runtime.Object) []string {
+	return nil
+}
+
+// WarningsOnUpdate returns warnings to the client performing the update.
+func (namespacedCloudProfileStrategy) WarningsOnUpdate(_ context.Context, _, _ runtime.Object) []string {
+	return nil
+}
+
+func dropExpiredVersions(namespacedCloudProfile *core.NamespacedCloudProfile) {
```

Comment:
Is it possible to make the original function reusable instead of duplicating it?

Reply:
Could surely be done but I don't see it as a big problem as the functions do differ from each other because the `kubernetes` field is optional in the `NamespacedCloudProfile`.